python/cuml/metrics/hinge_loss.pyx

```diff
@@ -0,0 +1,120 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import cupy as cp
+from cuml.preprocessing import LabelEncoder, LabelBinarizer
+import cudf
+
+
+def cython_hinge_loss(y_true, pred_decision, labels=None, sample_weights=None):
+    """
+    Calculates non-regularized hinge loss. Adapted from scikit-learn hinge loss
+
+    Parameters
+    ----------
+    y_true: cuDF Series or cuPy array of shape (n_samples,)
+            True labels, consisting of labels for the classes.
+            In binary classification, the positive label must be
+            greater than negative class
+
+    pred_decision: cuDF DataFrame or cuPy array of shape (n_samples,) or
+                   (n_samples, n_classes)
+                   Predicted decisions, as output by decision_function (floats)
+
+    labels: cuDF Series or cuPy array, default=None
+            In multiclass problems, this must include all class labels.
+
+    sample_weight: cupy array of shape (n_samples,), default=None
+                   Sample weights to be used for computing the average
+
+    Returns
+    -------
+    loss : float.
+           The average hinge loss.
+    """
+
+    # Check types of the inputs
+    if not hasattr(y_true, "__cuda_array_interface__"):
+        raise TypeError("y_true needs to be either a cuDF Series or \
+                        a cuda_array_interface compliant array.")
+
+    if not hasattr(pred_decision, "__cuda_array_interface__") and not \
+       isinstance(pred_decision, cudf.DataFrame):
+        raise TypeError("pred_decision needs to be either a cuDF DataFrame or \
+                        a cuda_array_interface compliant array.")
+
+    if y_true.shape[0] != pred_decision.shape[0]:
+        raise ValueError("y_true and pred_decision must have the same"
+                         " number of rows(found {} and {})".format(
+                             y_true.shape[0],
+                             pred_decision.shape[0]))
+
+    if sample_weights and sample_weights.shape[0] != y_true.shape[0]:
+        raise ValueError("y_true and sample_weights must have the same "
+                         "number of rows (found {} and {})".format(
+                             y_true.shape[0],
+                             sample_weights.shape[0]))
+
+    y_cudf = isinstance(y_true, cudf.Series)
+    labels_cudf = isinstance(labels, cudf.Series)
+
+    if not labels_cudf:
+        labels = cudf.Series(labels)
+
+    if not y_cudf:
+        y_true = cudf.Series(y_true)
+
+    if len(y_true.shape) != 1:
```

Comment:
If `y_true` is a `Series` (as ensured at line 78), then it should always be 1D, and this check can be avoided, right?

Reply:
You're right, I has initially included this above the conversion, which makes more sense. I'll move it up.